pkg/cmd/ssh/command.go

```diff
@@ -0,0 +1,801 @@
+	// connections on the SSHPort and has a public key configured that matches the
+	// given private key.
+	bastionAvailabilityChecker = func(hostname string, privateKey []byte) error {
+		authMethods := []ssh.AuthMethod{}
+
+		if len(privateKey) > 0 {
+			signer, err := ssh.ParsePrivateKey(privateKey)
+			if err != nil {
+				return fmt.Errorf("invalid private SSH key: %w", err)
+			}
+
+			authMethods = append(authMethods, ssh.PublicKeys(signer))
+		} else if addr := os.Getenv("SSH_AUTH_SOCK"); len(addr) > 0 {
+			socket, dialErr := net.Dial("unix", addr)
+			if dialErr != nil {
+				return fmt.Errorf("could not open SSH agent socket %q: %w", addr, dialErr)
+			}
+			defer socket.Close()
+
+			signers, signersErr := agent.NewClient(socket).Signers()
+			if signersErr != nil {
+				return fmt.Errorf("error when creating signer for SSH agent: %w", signersErr)
+			}
+
+			authMethods = append(authMethods, ssh.PublicKeys(signers...))
+		} else {
+			return errors.New("neither private key nor the environment variable SSH_AUTH_SOCK are defined, cannot connect to bastion")
+		}
+
+		client, err := ssh.Dial("tcp", net.JoinHostPort(hostname, strconv.Itoa(SSHPort)), &ssh.ClientConfig{
+			User:            SSHBastionUsername,
+			HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+			Auth:            authMethods,
+			Timeout:         10 * time.Second,
+		})
+
+		if client != nil {
+			client.Close()
+		}
+
+		return err
+	}
+
+	// bastionNameProvider generates the name for a new bastion.
+	bastionNameProvider = func() (string, error) {
+		bastionID, err := utils.GenerateRandomString(8)
+		if err != nil {
+			return "", err
+		}
+
+		return fmt.Sprintf("cli-%s", strings.ToLower(bastionID)), nil
+	}
+
+	// createSignalChannel returns a channel which receives OS signals.
+	createSignalChannel = func() chan os.Signal {
+		signalChan := make(chan os.Signal, 1)
+		signal.Notify(signalChan, os.Interrupt)
+
+		return signalChan
+	}
+
+	// execCommand executes the given command, using the in/out streams
+	// from the options. The function returns an error if the command
+	// fails.
+	execCommand = func(ctx context.Context, command string, args []string, o *Options) error {
+		cmd := exec.CommandContext(ctx, command, args...)
+		cmd.Stdout = o.IOStreams.Out
+		cmd.Stdin = o.IOStreams.In
+		cmd.Stderr = o.IOStreams.ErrOut
+
+		return cmd.Run()
+	}
+)
+
+// NewCommand returns a new ssh command.
+func NewCommand(f util.Factory, o *Options) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "ssh [NODE_NAME]",
+		Short: "Establish an SSH connection to a Shoot cluster's node",
+		Args:  cobra.MaximumNArgs(1),
+		ValidArgsFunction: func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+			if len(args) != 0 {
+				return nil, cobra.ShellCompDirectiveNoFileComp
+			}
+
+			nodeNames, err := getNodeNamesFromShoot(f, toComplete)
+			if err != nil {
+				fmt.Fprintln(o.IOStreams.ErrOut, err.Error())
```

Comment:
for debugging purposes, cobra suggests to use https://github.com/spf13/cobra/blob/master/shell_completions.md#dynamic-completion-of-nouns

```go
// Prints to the completion script debug file (if BASH_COMP_DEBUG_FILE
// is set to a file path) and optionally prints to stderr.
cobra.CompDebug(msg string, printToStdErr bool) {
cobra.CompDebugln(msg string, printToStdErr bool)

// Prints to the completion script debug file (if BASH_COMP_DEBUG_FILE
// is set to a file path) and to stderr.
cobra.CompError(msg string)
cobra.CompErrorln(msg string)
```

however you can't pass your own IOStream unfortunately.

Reply:
So what sacrifice do we want to make here? ^^

Reply:
I give you the honour to choose :) I do not have a strong opinion here

Reply:
Then I would simply keep it as it. The "convenience" of replacing a single `fmt.Println()` call with something else is not worth losing the abstraction that is IOStreams.

Reply:
Fine 👍